presto-spark-base/src/main/java/com/facebook/presto/spark/execution/AbstractPrestoSparkQueryExecution.java

```diff
@@ -708,6 +717,144 @@ else if (writerTarget instanceof ExecutionWriterTarget.RefreshMaterializedViewHa
         }
     }
 
+    // Returns RDD for specified fragmented subplan
+    // This method ensures that RDD is created only once for a sub-plan, where identity is determined by fragment id
+    protected <T extends PrestoSparkTaskOutput> RddAndMore<T> createRddFragmented(SubPlan subPlan, TableWriteInfo tableWriteInfo)
+            throws SparkException, TimeoutException
+    {
+        synchronized (fragmentIdToRdd) {
```

Comment:
just make the whole method synchronized.  Synchronizing on the map doesn't help us.

Reply:
Done. IMO Synchronizing the map is keeps locking limited to this map(which we need), rather than locking whole object.

Reply:
Synchronizing on an object doesn't lock the object, it uses the object as a lock, which is different.  e.g. Someone could still modify that map at the same time if they aren't using a synchronized access. but anything synchronized on the same object will need to acquire the same lock.  the lock you synchronize on doesn't even need to be related to the object(s) you are trying to modify.  You can have a String called "lock" and synchronize on that.  That's where the `@guardedBy` annotation comes in.  It tells you what lock you need to acquire to safely use that object.

IMO the only reason to lock on something other than the class for basic synchronization, is if you need more than one lock.

Reply:
"Synchronizing on an object doesn't lock the object, it uses the object as a lock" - yeah, I meant acquiring the lock on the object as there is no "locking" object. In this case, we want to ensure that map is accessed in synchronized manner. I was going for fine grained locking. 

"IMO the only reason to lock on something other than the class for basic synchronization, is if you need more than one lock." -> Exactly, I was going for this futuristic scenario. Suppose, we end up introducing another method which requires synchronized behavior. Keeping the lock at AbstractPrestoSparkQueryExecution will mean that these two methods can't run at same time. Which may or may not be what we want. I was keeping the intention of this method clear that it needs locking on just the map.

For now, it's fine to synchronized this method. I will update the PR